google/cloud/pubsublite/internal/partition_publisher.cc

```diff
@@ -0,0 +1,353 @@
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+
+using google::cloud::pubsublite::v1::Cursor;
+using google::cloud::pubsublite::v1::InitialPublishRequest;
+using google::cloud::pubsublite::v1::MessagePublishRequest;
+using google::cloud::pubsublite::v1::PublishRequest;
+using google::cloud::pubsublite::v1::PublishResponse;
+using google::cloud::pubsublite::v1::PubSubMessage;
+
+using UnderlyingStream = std::unique_ptr<
+    AsyncStreamingReadWriteRpc<google::cloud::pubsublite::v1::PublishRequest,
+                               google::cloud::pubsublite::v1::PublishResponse>>;
+
+PartitionPublisherImpl::PartitionPublisherImpl(
+    absl::FunctionRef<std::unique_ptr<
+        ResumableAsyncStreamingReadWriteRpc<PublishRequest, PublishResponse>>(
+        StreamInitializer<PublishRequest, PublishResponse>)>
+        resumable_stream_factory,
+    BatchingSettings const& batching_settings, InitialPublishRequest const& ipr,
+    std::unique_ptr<AlarmRegistryInterface> alarm, size_t alarm_period_ms)
+    : batching_settings_{batching_settings},
+      ipr_{ipr},
+      resumable_stream_{resumable_stream_factory(std::bind(
+          &PartitionPublisherImpl::Initializer, this, std::placeholders::_1))},
+      cancel_token_{alarm->RegisterAlarm(
+          absl::Milliseconds(alarm_period_ms),
+          std::bind(&PartitionPublisherImpl::OnAlarm, this))} {}
+
+PartitionPublisherImpl::~PartitionPublisherImpl() {
+  future<void> shutdown = Shutdown();
+  if (!shutdown.is_ready()) {
+    GCP_LOG(WARNING) << "`Shutdown` must be called and finished before object "
+                        "goes out of scope.";
+    assert(false);
+  }
+  shutdown.get();
+}
+
+future<Status> PartitionPublisherImpl::Start() {
+  promise<void> root;
+  future<void> root_future = root.get_future();
+  future<Status> return_future;
+  {
+    std::lock_guard<std::mutex> g{mu_};
+    root_future.then(ChainFuture(resumable_stream_->Start()))
+        .then([this](future<Status> start_status) {
+          bool has_shutdown = true;
+          promise<Status> start_promise{null_promise_t{}};
+          {
+            std::lock_guard<std::mutex> g{mu_};
+            if (!start_future_) return;
+            if (!shutdown_) {
+              has_shutdown = false;
+              shutdown_ = true;
+            }
+            start_promise = std::move(*start_future_);
+            start_future_.reset();
+          }
+          Status status = start_status.get();
+          start_promise.set_value(status);
+          if (!has_shutdown) {
+            cancel_token_ = nullptr;
+            SatisfyOutstandingMessages(status);
+          }
+        });
+  }
+  root.set_value();
+  Read();
+  return start_future_->get_future();
+}
+
+future<StatusOr<Cursor>> PartitionPublisherImpl::Publish(PubSubMessage m) {
+  future<StatusOr<Cursor>> message_future;
+  {
+    std::lock_guard<std::mutex> g{mu_};
+    if (shutdown_) return make_ready_future(StatusOr<Cursor>(Status()));
```

Comment:
this will CHECK-fail. A StatusOr may never hold an OK status. Create an AbortedError with "Already shut down." as the message